### PR TITLE
[GLUTEN-1868][VL] Optimize ColumnarToRow to reuse the allocated Buffer in native side.

### PR DESCRIPTION
**Pull request by JkSelf**
## What changes were proposed in this pull request?

This PR make all the ColumnarBatch in one partition to use same buffer as much as possible.

https://github.com/oap-project/gluten/issues/1868

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala [hunk 1]
```diff
@@ -133,73 +133,90 @@ class ColumnarToRowRDD(@transient sc: SparkContext, rdd: RDD[ColumnarBatch],
...
+            val batchHandle = GlutenColumnarBatches.getNativeHandle(offloaded)
+
+            if (c2rId == -1) {
+              c2rId = jniWrapper.nativeColumnarToRowInit(
+                batchHandle,
+                NativeMemoryAllocators.contextInstance().getNativeInstanceId)
+            }
+            val info = jniWrapper.nativeColumnarToRowWrite(batchHandle, c2rId)
+
+            convertTime += NANOSECONDS.toMillis(System.nanoTime() - beforeConvert)
```
Comment (jinchengchenghh) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/VeloxColumnarToRowExec.scala:196]: NANOSECONDS.toMillis(System.nanoTime() - beforeConvert) -> System.currenttimeMillis()

File: cpp/core/jni/JniWrapper.cc [hunk 2]
```diff
@@ -541,16 +540,26 @@ Java_io_glutenproject_vectorized_NativeColumnarToRowJniWrapper_nativeConvertColu
...
+  JNI_METHOD_START
+  auto columnarToRowConverter = columnarToRowConverterHolder.lookup(instanceId);
+  std::shared_ptr<ColumnarBatch> cb = glutenColumnarbatchHolder.lookup(batchHandle);
+  GLUTEN_THROW_NOT_OK(columnarToRowConverter->write(cb));
 
   const auto& offsets = columnarToRowConverter->getOffsets();
   const auto& lengths = columnarToRowConverter->getLengths();
-  int64_t instanceID = columnarToRowConverterHolder.insert(columnarToRowConverter);
+
+  auto numRows = cb->getNumRows();
```
Comment (jinchengchenghh) [inline cpp/core/jni/JniWrapper.cc:562]: Where to use `numRows`?

Comment (JkSelf) [reply cpp/core/jni/JniWrapper.cc:562]: https://github.com/JkSelf/gluten/blob/9e0edbedff2a67146b98b41e3cff971c215b247c/cpp/core/jni/JniWrapper.cc#L566  and https://github.com/JkSelf/gluten/blob/9e0edbedff2a67146b98b41e3cff971c215b247c/cpp/core/jni/JniWrapper.cc#L569 use the `numRows`.

**Comment by JkSelf** [comment]
@jinchengchenghh Please help to review again. Thanks.

**Review by jinchengchenghh — APPROVED** [review]

**Pull request merged**